tests/v2/validation/observability/stackstate_test.go

```diff
@@ -0,0 +1,338 @@
+}
+
+func (ss *StackStateTestSuite) TearDownSuite() {
+	ss.session.Cleanup()
+}
+
+func (ss *StackStateTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	ss.session = testSession
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(ss.T(), err)
+
+	ss.client = client
+
+	clusterName := client.RancherConfig.ClusterName
+	require.NotEmptyf(ss.T(), clusterName, "Cluster name to install should be set")
+	cluster, err := clusters.NewClusterMeta(ss.client, clusterName)
+	require.NoError(ss.T(), err)
+	ss.cluster = cluster
+
+	log.Info("Creating a project and namespace for the chart to be installed in.")
+
+	projectTemplate := kubeprojects.NewProjectTemplate(cluster.ID)
+	projectTemplate.Name = charts.StackstateNamespace
+	project, err := client.Steve.SteveType(project).Create(projectTemplate)
+	require.NoError(ss.T(), err)
+	ss.projectID = project.ID
+
+	_, err = namespaces.CreateNamespace(client, cluster.ID, project.Name, charts.StackstateNamespace, "", map[string]string{}, map[string]string{})
+	require.NoError(ss.T(), err)
+
+	_, err = ss.client.Catalog.ClusterRepos().Get(context.TODO(), rancherUIPlugins, meta.GetOptions{})
+
+	if k8sErrors.IsNotFound(err) {
+		err = observability.CreateExtensionsRepo(ss.client, rancherUIPlugins, uiExtensionsRepo, uiGitBranch)
+		log.Info("Created an extensions repo for ui plugins.")
+	}
+	require.NoError(ss.T(), err)
+
+	var stackstateConfigs observability.StackStateConfigs
+	config.LoadConfig(stackStateConfigFileKey, &stackstateConfigs)
+	ss.stackstateConfigs = stackstateConfigs
+
+	err = observability.WhitelistStackstateDomains(ss.client, []string{ss.stackstateConfigs.Url})
+	require.NoError(ss.T(), err)
+	log.Info("Node driver installed with stackstate extensions ui to whitelist stackstate URL")
+
+	crdsExists, err := ss.client.Steve.SteveType(observability.ApiExtenisonsCRD).ByID(observability.ObservabilitySteveType)
+	if crdsExists == nil && strings.Contains(err.Error(), "Not Found") {
+		err = observability.InstallStackstateCRD(ss.client)
+		log.Info("Installed stackstate CRD")
+	}
+	require.NoError(ss.T(), err)
+
+
+	client, err = client.ReLogin()
+	require.NoError(ss.T(), err)
+
+	initialStackstateExtension, err := extencharts.GetChartStatus(client, localCluster, charts.StackstateExtensionNamespace, charts.StackstateExtensionsName)
+	require.NoError(ss.T(), err)
+
+	if !initialStackstateExtension.IsAlreadyInstalled {
+		latestUIPluginVersion, err := ss.client.Catalog.GetLatestChartVersion(charts.StackstateExtensionsName, charts.UIPluginName)
+		require.NoError(ss.T(), err)
+
+		extensionOptions := &uiplugins.ExtensionOptions{
+			ChartName:   charts.StackstateExtensionsName,
+			ReleaseName: charts.StackstateExtensionsName,
+			Version:     latestUIPluginVersion,
+		}
+
+		err = uiplugins.InstallStackstateUiPlugin(client, extensionOptions)
+		require.NoError(ss.T(), err)
+		ss.T().Log("Installed stackstate ui extensions")
+	}
```

Comment:
the more I think about it, the more conflicted I am about having these conditionals in for installing the resources. I feel like it would be a better testing practice to either expect these prerequisites to be installed, or not installed, in their entirety.  

technically, these tests are just to install stackstate agent. so the if conditions here would be fine. But, that means we should have 1 end-to-end test where we actually depend on nothing being installed when the test starts, and verifying that stackstate can be installed from beginning to end.

Thoughts?

Reply:
There are somere challenges with that approach. We might work with setups where some resources were already installed, either due to previous installs. Relying on a fully unconfigured setup could lead to test failures if we don’t get a clean environment every time. 

These conditional statements allows that flexibility, I think having these checks aligns better with the various test steps. Otherwise we expect the user to come with a clean setup everytime they start these checks.